### ISSUE TITLE
dev Client/Server deadlock processing close operation

### ISSUE THREAD

**Issue by franke-hub** [franke-hub/Cornucopia#3]
Client and Server inp_task close processing deadlocks because inp_task invokes close which removes the last shared_ptr<Server> reference. Even adding a "keep_alive" shared_ptr in task_inp doesn't help enough because the Client/Server destructor is still invoked under task_inp control *before* it returns control to the Dispatcher. The deadlock occurs since in Dispatch::Task::~Task a wait operation is enqueued and waited for while task_inp is running, and the wait operation processing can't begin until task_inp returns control to the Dispatcher.
 
FIX: In inp_task, we defer posting of the CLOSE work Item by transferring close work Items onto an internal Dispatcher task (which was created for this purpose.) When that Task posts the CLOSE work item, the last Client/Server shared pointer reference is removed. Task destruction runs under the internal Dispatcher Task, thus removing the deadlock.

Note: This fix will be committed to the maint branch, which still contains extensive debugging code. When testing successfully completes this issue will be closed, the extensive debugging code will be removed, and the maint branch merged into the trunk.

**Comment by franke-hub** [comment]
Regression test completed.

**Comment by franke-hub** [comment]
Merged into branches trunk and maint
